src/main/java/org/opentripplanner/middleware/tripmonitor/TrustedCompanion.java

```diff
@@ -0,0 +1,213 @@
+        Locale locale = I18nUtils.getOtpUserLocale(relatedUser);
+
+        String acceptDependentLinkLabel = Message.ACCEPT_DEPENDENT_EMAIL_LINK_TEXT.get(locale);
+        String acceptDependentUrl = getAcceptDependentUrl(acceptKey);
+        String addressee = (dependentUser.name != null) ? dependentUser.name : dependentUser.email;
+
+        // A HashMap is needed instead of a Map for template data to be serialized to the template renderer.
+        Map<String, Object> templateData = new HashMap<>(
+            Map.of(
+                "acceptDependentLinkAnchorLabel", acceptDependentLinkLabel,
+                "acceptDependentLinkLabelAndUrl", label(acceptDependentLinkLabel, acceptDependentUrl, locale),
+                "acceptDependentUrl", acceptDependentUrl,
+                "emailFooter", Message.ACCEPT_DEPENDENT_EMAIL_FOOTER.get(locale),
+                "emailGreeting", String.format(Message.ACCEPT_DEPENDENT_EMAIL_GREETING.get(locale), addressee),
+                "manageLinkUrl", String.format("%s%s", OTP_UI_URL, SETTINGS_PATH),
+                "manageLinkText", Message.ACCEPT_DEPENDENT_EMAIL_MANAGE.get(locale)
+            )
+        );
+
+        return NotificationUtils.sendEmail(
+            relatedUser,
+            Message.ACCEPT_DEPENDENT_EMAIL_SUBJECT.get(locale),
+            "AcceptDependentText.ftl",
+            "AcceptDependentHtml.ftl",
+            templateData
+        );
+    }
+
+    private static String getAcceptDependentUrl(String acceptKey) {
+        return String.format("%s/%s%s", AWS_API_SERVER, AWS_API_STAGE, getAcceptDependentEndPoint(acceptKey));
+    }
+
+    public static String getAcceptDependentEndPoint(String acceptKey) {
+        return String.format("/%s?%s=%s", ACCEPT_DEPENDENT_PATH, ACCEPT_KEY, acceptKey);
+    }
+
+    /**
+     * @return the {@link OtpUser} found with a {@link RelatedUser#acceptKey} in {@link OtpUser#relatedUsers} that
+     * matches the provided acceptKey.
+     */
+    private static OtpUser getUserForAcceptKey(String acceptKey) {
+        return Persistence.otpUsers.getOneFiltered(Filters.elemMatch("relatedUsers", Filters.eq(ACCEPT_KEY, acceptKey)));
+    }
+
+    /**
+     * If a dependent removes a related user, remove the dependent from the related user.
+     */
+    public static void ensureRelatedUserIntegrity(OtpUser updatedUser, OtpUser preExistingUser) {
+        List<RelatedUser> difference = preExistingUser.relatedUsers
+            .stream()
+            .filter(relatedUser -> !updatedUser.relatedUsers.contains(relatedUser))
+            .collect(Collectors.toList());
+        for (RelatedUser relatedUser : difference) {
+            OtpUser user = Persistence.otpUsers.getOneFiltered(eq(EMAIL_FIELD_NAME, relatedUser.email));
+            if (user != null) {
+                user.dependents.remove(updatedUser.id);
+                Persistence.otpUsers.replace(user.id, user);
+            }
+        }
```

Comment:
Reuse this `for` block in `OtpUser::delete`.

Reply:
Update in: https://github.com/ibi-group/otp-middleware/pull/256/commits/99341cb2a25e6633d69fb424482c7936fb6adccc